### ISSUE TITLE
multi-allelic site issues (2)

### ISSUE THREAD

**Issue by MatthewMaher** [opencb/opencga#503]
My apologies if this is too long, but I do not think the issue can be simply described. Skip down to 'Issues/Question' to see the most concise questions. I believe all these issues/questions stem from having a multi-allelic variants.

At chr6  position 10763779 the reference has "TAAA" and there is a wide variation in the # of As that get deleted there.

In the VCFs:

for 'sample1' (aka **file 4**) the VCF is bi-allelic (delete one A and delete two As):
```
chr6    10763779    .    TAA    T,TA    500.73    .    AC=1,1;AF=0.500,0.500;AN=2;DP=28;ExcessHet=3.0103;FS=0.000;MLEAC=1,1;MLEAF=0.500,0.500;MQ=60.00;QD=23.84;SOR=3.816    GT:AD:DP:GQ:PL    1/2:0,9,12:21:99:538,258,250,215,0,156
```
for 'sample2' (HOM for delete one A):
```
chr6    10763779    .    TA    T    222.73    .    AC=2;AF=1.00;AN=2;BaseQRankSum=-1.044;ClippingRankSum=-1.741;DP=23;ExcessHet=3.0103;FS=0.000;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=1.393;QD=17.13;ReadPosRankSum=-1.741;SOR=1.802    GT:AD:DP:GQ:PL    1/1:1,12:13:15:225,15,0
```
for 'sample3' (HET for delete **three** As):
```
chr6    10763779    .    TAAA    T    99.73    .    AC=1;AF=0.500;AN=2;DP=7;ExcessHet=3.0103;FS=0.000;MLEAC=1;MLEAF=0.500;MQ=60.00;QD=19.95;SOR=0.693    GT:AD:DP:GQ:PL    0/1:2,3:5:6:86,0,6
```

When openCGA loads these samples, three relevant things seem to occur:

1:  their representation strips off the common first base (the 'T'), so the position gets bumped by one to 10763780.  Sounds good to me.
2:  that first bi-allelic VCF row that contains multiple HET calls gets deconstructed into two variant entries in openCGA.
3:  calls from all samples get merged into a single unified data representation, so openCGA essentially knows that this is a TRI-allelic site, and attempts to report it as such.

When we want to retrieve data for a specific sample (file 4, sample1), we are doing it with a URL like this (except I've added a position range to focus on this location):
```
<baseurl>/opencga/webservices/rest/v1/analysis/variant/query?sid=<sid>&chromosome=6%3A10763780-10763780&studies=2&returnedSamples=sample1&files=4&returnedFiles=4
```

## Bug/Problems?

 - The attributes that are being displayed as coming from file 4 in **Studies->Files** are NOT from file 4.

 - The data within **studies->SamplesData** (specifically **AD** and **PL**) do not get properly rearranged (and enlarged!) to match the properly adjusted GT values. They all (GT,AD,PL) need adjusting because their original values are single-file-specific, but extra alternate alleles, seen in additional files, are being returned in the output.  I believe the **GT** values look correctly adjusted, but not the **AD,PL**. 

## Related Question

Our intent is to retrieve a non-duplicate list of all the variants for one sample.  OpenCGA returns each allele as a separate variant, and also supplies all the alternates (seen in ANY file) under **secondaryAlternates** and the GT field appears to be recalculated based on those primary and secondary alternates.  My understanding is that all annotations such as **consequenceType**, pertain to the *primary* alternate only (i.e. the GT=1 case).  So I *believe* that the way to get a complete, annotated and non-duplicate list of variants for a given sample, is to process each variant, but only consider the instances where the Studies->Files->GT contains at least one '1'.  **Is that correct?**   (so in the example below, I would ignore the last variant where GT='2/3')

#### look for (...) comments on right side
sample below has been edited for brevity
```
{
    "apiVersion": "v1",
    "warning": "",
    "error": "",
    "queryOptions": {
        "metadata": true,
        "limit": 2000
    },
    "response": [
        {
            "time": 0,
            "dbTime": 65,
            "numResults": 3,
            "numTotalResults": -1,
            "resultType": "org.opencb.biodata.models.variant.Variant",
            "result": [
                {
                    "id": "6:10763780:A:-",
                    "names": [],
                    "chromosome": "6",
                    "start": 10763780,
                    "strand": "+",
                    "reference": "A",
                    "alternate": "",
                    "studies": [
                        {
                            "files": [
                                {
                                    "fileId": "4",  (all attributes below are from the wrong file)
                                    "call": "10763779:TA:T:0",
                                    "attributes": {
                                        "AC": "2",
                                        "MQRankSum": "1.393",
                                        "FILTER": ".",
                                        "MQ": "60.0",
                                        "AF": "1.0",
                                        "MLEAC": "1",
                                        "BaseQRankSum": "-1.044",
                                        "ExcessHet": "3.0103",
                                        "QUAL": "222.73",
                                        "MLEAF": "0.5",
                                        "DP": "23",
                                        "ReadPosRankSum": "-1.741",
                                        "AN": "2",
                                        "FS": "0.0",
                                        "QD": "17.13",
                                        "SOR": "1.802",
                                        "ClippingRankSum": "-1.741"
                                    }
                                }
                            ],
                            "secondaryAlternates": [
                                {
                                    "chromosome": "6",
                                    "start": 10763780,
                                    "end": 10763781,
                                    "reference": "AA",
                                    "alternate": "",
                                    "type": "INDEL"
                                },
                                {
                                    "chromosome": "6",
                                    "start": 10763780,
                                    "end": 10763782,
                                    "reference": "AAA",
                                    "alternate": "",
                                    "type": "INDEL"
                                }
                            ],
                            "samplesData": [
                                [
                                    "1/2",
                                    "0,9,12",  (I think this should be 0,12,9,0 )
                                    "21",
                                    "99",
                                    "538,215,156,258,0,250"  (should have 10 values)
                                ]
                            ],
                            "format": [
                                "GT",
                                "AD",
                                "DP",
                                "GQ",
                                "PL"
                            ]
                        }
                    ],
                    "type": "INDEL",
                    "end": 10763780,
                    "length": 1
                },
                {
                    "id": "6:10763780:AA:-",
                    "names": [],
                    "chromosome": "6",
                    "start": 10763780,
                    "strand": "+",
                    "reference": "AA",
                    "alternate": "",
                    "studies": [
                        {
                            "files": [
                                {
                                    "fileId": "4",   (all attributes below are from the wrong file)
                                    "call": "10763779:TA:T:0",
                                    "attributes": {
                                        "AC": "2",
                                        "MQRankSum": "1.393",
                                        "FILTER": ".",
                                        "MQ": "60.0",
                                        "AF": "1.0",
                                        "MLEAC": "1",
                                        "BaseQRankSum": "-1.044",
                                        "ExcessHet": "3.0103",
                                        "QUAL": "222.73",
                                        "MLEAF": "0.5",
                                        "DP": "23",
                                        "ReadPosRankSum": "-1.741",
                                        "AN": "2",
                                        "FS": "0.0",
                                        "QD": "17.13",
                                        "SOR": "1.802",
                                        "ClippingRankSum": "-1.741"
                                    }
                                }
                            ],
                            "secondaryAlternates": [
                                {
                                    "chromosome": "6",
                                    "start": 10763780,
                                    "end": 10763780,
                                    "reference": "A",
                                    "alternate": "",
                                    "type": "INDEL"
                                },
                                {
                                    "chromosome": "6",
                                    "start": 10763780,
                                    "end": 10763782,
                                    "reference": "AAA",
                                    "alternate": "",
                                    "type": "INDEL"
                                }
                            ],
                            "samplesData": [
                                [
                                    "1/2",
                                    "0,9,12",   (I think this should be 0,9,12,0 )
                                    "21",
                                    "99",
                                    "538,258,250,215,0,156"   (should have 10 values)
                                ]
                            ],
                            "format": [
                                "GT",
                                "AD",
                                "DP",
                                "GQ",
                                "PL"
                            ]
                        }
                    ],
                    "type": "INDEL",
                    "end": 10763781,
                    "length": 2
                },
                {
                    "id": "6:10763780:AAA:-",
                    "names": [],
                    "chromosome": "6",
                    "start": 10763780,
                    "strand": "+",
                    "reference": "AAA",
                    "alternate": "",
                    "studies": [
                        {
                            "files": [
                                {
                                    "fileId": "4",    (all attributes below are from the wrong file)
                                    "call": "10763779:TA:T:0",
                                    "attributes": {
                                        "AC": "2",
                                        "MQRankSum": "1.393",
                                        "FILTER": ".",
                                        "MQ": "60.0",
                                        "AF": "1.0",
                                        "MLEAC": "1",
                                        "BaseQRankSum": "-1.044",
                                        "ExcessHet": "3.0103",
                                        "QUAL": "222.73",
                                        "MLEAF": "0.5",
                                        "DP": "23",
                                        "ReadPosRankSum": "-1.741",
                                        "AN": "2",
                                        "FS": "0.0",
                                        "QD": "17.13",
                                        "SOR": "1.802",
                                        "ClippingRankSum": "-1.741"
                                    }
                                }
                            ],
                            "secondaryAlternates": [
                                {
                                    "chromosome": "6",
                                    "start": 10763780,
                                    "end": 10763780,
                                    "reference": "A",
                                    "alternate": "",
                                    "type": "INDEL"
                                },
                                {
                                    "chromosome": "6",
                                    "start": 10763780,
                                    "end": 10763781,
                                    "reference": "AA",
                                    "alternate": "",
                                    "type": "INDEL"
                                }
                            ],
                            "samplesData": [
                                [
                                    "2/3",      (properly adjusted)
                                    "0,9,12", (I think this should be 0,0,12,9 )
                                    "21",
                                    "99",
                                    "538,215,156,258,0,250"  (should have 10 values)
                                ]
                            ],
                            "format": [
                                "GT",
                                "AD",
                                "DP",
                                "GQ",
                                "PL"
                            ]
                        }
                    ],
                    "type": "INDEL",
                    "end": 10763782,
                    "length": 3
                }
            ]
        }
    ]
}
```

**Comment by MatthewMaher** [comment]
My apologies, but I now suspect that the first of the two problems is not real - as I'm unable to recreate it now - I may have gotten confused by my team member reloading the files and the file #s for each sample getting changed.    Please ignore that first issue about the data associated with "File 4" seeming to be from the wrong file.  

Please only investigate the second issue pertaining to "The data within studies->SamplesData (specifically AD and PL)".  Hopefully it is clear that the these multi-valued fields are positionally related to the various alternate alleles.  But since additional alternate alleles are being added from what was in the original VCF input, then these fields should also require some recalculating.  

Sorry for any confusion.
-Matt

**Comment by j-coll** [comment]
Hi @MatthewMaher ,

Thanks for the detailed post.

The three things that you have noticed are right. As you noticed, the idea of separating the variant into different variant entries is for having specific annotation, since a VariantAnnotation only refers to one genomic event, i.e. the _primary_ alternate. Normalizing the data helps us to relate variants from different samples in the same study/dataset. For now, you'll need to filter the variants that contains at least one '1'. In parallel, we are working in a new set of filters: #494. See the filter "**SAMPLES**"

Either it is a problem with the code or not, the fact that you got confused about the file ids, is a problem. In the past we had some small aliases (user defined) to write instead of the numerical id. The idea of writing the whole file name was discarded to avoid incrementing the size of the returned data.
We are preparing some rest endpoints with samples and files metadata for variant queries.

About the second problem, it's true that there are some fields not rearranged. It has been a pending issue for to much time now. I will create a new issue referring to this problem.
